funsor/distribution.py

```diff
@@ -57,12 +57,34 @@ class DistributionMeta(FunsorMeta):
     """
     def __call__(cls, *args, **kwargs):
         kwargs.update(zip(cls._ast_fields, args))
-        value = kwargs.pop('value', 'value')
-        kwargs = OrderedDict(
-            (k, to_funsor(kwargs[k], output=cls._infer_param_domain(k, getattr(kwargs[k], "shape", ()))))
-            for k in cls._ast_fields if k != 'value')
-        value = to_funsor(value, output=cls._infer_value_domain(**{k: v.output for k, v in kwargs.items()}))
-        args = numbers_to_tensors(*(tuple(kwargs.values()) + (value,)))
+        kwargs["value"] = kwargs.get("value", "value")
+        kwargs = OrderedDict((k, kwargs[k]) for k in cls._ast_fields)  # make sure args are sorted
+
+        domains = OrderedDict()
+        for k, v in kwargs.items():
+            if k == "value":
+                continue
+
+            # compute unbroadcasted param domains
+            domain = cls._infer_param_domain(k, getattr(kwargs[k], "shape", ()))
+            # use to_funsor to infer output dimensions of e.g. tensors
+            domains[k] = domain if domain is not None else to_funsor(v).output
+
+            # broadcast individual param domains with Funsor inputs
+            # this avoids .expand-ing underlying parameter tensors
```

Comment:
What is the expected domain of `scale` for `Normal(Reals[2], 1.)` and `Normal(Reals[2], torch.ones(2))`. Currently, `domains["scale"]` will be `Real` in both case. The second case will trigger an error at `to_funsor(v, output=domains[k])` below.

In either case, I guess we need to rewrite `eager_normal` or `eager_mvn` to address `Reals[2]` loc. cc @fritzo

Reply:
> What is the expected domain of scale for `Normal(Reals[2], 1.)` and `Normal(Reals[2], torch.ones(2))`?

In the first case, it's `Real`, and in the second, it's `Reals[2]`.  I guess I should add a second broadcasting condition below to handle the case where the parameter is a raw tensor:
```py
if ops.is_numeric_array(v):  # at this point we know all of v's dims are output dims
    domains[k] = Reals[broadcast_shape(v.shape, domains[k].shape)]
```

> In either case, I guess we need to rewrite eager_normal or eager_mvn to address Reals[2] loc

Can you explain what change you think may be necessary?